### ISSUE TITLE
IndexOutOfBoundsException from ZeroCopyDataBlockSerde

### ISSUE THREAD

**Issue by Jackie-Jiang** [apache/pinot#14375]
```
java.lang.AssertionError: 
Caught exception while testing query!
Pinot query: SELECT "DivArrDelay", "Cancelled", "DestAirportID" FROM mytable WHERE "OriginStateName" BETWEEN 'Montana' AND 'South Dakota' AND "OriginAirportID" BETWEEN 13127 AND 12945 OR "DistanceGroup" = 4 ORDER BY "Month", "LateAircraftDelay", "TailNum" LIMIT 10000
H2 query: SELECT `DivArrDelay`, `Cancelled`, `DestAirportID` FROM mytable WHERE `OriginStateName` BETWEEN 'Montana' AND 'South Dakota' AND `OriginAirportID` BETWEEN 13127 AND 12945 OR `DistanceGroup` = 4 ORDER BY `Month`, `LateAircraftDelay`, `TailNum` LIMIT 10000
	at org.testng.Assert.fail(Assert.java:99)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.failure(ClusterIntegrationTestUtils.java:1081)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.failure(ClusterIntegrationTestUtils.java:1064)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQuery(ClusterIntegrationTestUtils.java:702)
	at org.apache.pinot.integration.tests.BaseClusterIntegrationTest.testQuery(BaseClusterIntegrationTest.java:750)
	at org.apache.pinot.integration.tests.BaseClusterIntegrationTestSet.testGeneratedQueries(BaseClusterIntegrationTestSet.java:494)
	at org.apache.pinot.integration.tests.MultiStageEngineIntegrationTest.testGeneratedQueries(MultiStageEngineIntegrationTest.java:175)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:141)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:686)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:230)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:63)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:992)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:203)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:154)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:134)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.testng.TestRunner.privateRun(TestRunner.java:739)
	at org.testng.TestRunner.run(TestRunner.java:614)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:421)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:413)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:373)
	at org.testng.SuiteRunner.run(SuiteRunner.java:312)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1274)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1208)
	at org.testng.TestNG.runSuites(TestNG.java:1112)
	at org.testng.TestNG.run(TestNG.java:1079)
	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:155)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.executeSingleClass(TestNGDirectoryTestSuite.java:102)
	at org.apache.maven.surefire.testng.TestNGDirectoryTestSuite.execute(TestNGDirectoryTestSuite.java:91)
	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:137)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: java.lang.RuntimeException: Got Exceptions from Query Response: {"numRowsResultSet":0,"partialResult":true,"exceptions":[{"errorCode":200,"message":"QueryExecutionError:\nReceived error query execution result block: {1000=Caught exception while processing blocks for mailbox: 2087751583000000048|1|0|0|0\njava.lang.RuntimeException: Caught exception while processing blocks for mailbox: 2087751583000000048|1|0|0|0\n\tat org.apache.pinot.query.mailbox.channel.MailboxContentObserver.onNext(MailboxContentObserver.java:95)\n\tat org.apache.pinot.query.mailbox.channel.MailboxContentObserver.onNext(MailboxContentObserver.java:41)\n\tat io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onMessage(ServerCalls.java:262)\n\tat io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.messagesAvailableInternal(ServerCallImpl.java:334)\n...\nCaused by: java.io.IOException: Failed to deserialize data block with serde class org.apache.pinot.common.datablock.ZeroCopyDataBlockSerde\n\tat org.apache.pinot.common.datablock.DataBlockUtils.deserialize(DataBlockUtils.java:251)\n\tat org.apache.pinot.common.datablock.DataBlockUtils.readFrom(DataBlockUtils.java:179)\n\tat org.apache.pinot.query.mailbox.ReceivingMailbox.offerRaw(ReceivingMailbox.java:108)\n\tat org.apache.pinot.query.mailbox.channel.MailboxContentObserver.onNext(MailboxContentObserver.java:63)\n...\nCaused by: java.lang.IndexOutOfBoundsException\n\tat java.base/java.nio.Buffer$1.apply(Buffer.java:757)\n\tat java.base/java.nio.Buffer$1.apply(Buffer.java:754)\n\tat java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:213)\n\tat java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:210)}\norg.apache.pinot.query.service.dispatch.QueryDispatcher.runReducer(QueryDispatcher.java:482)\norg.apache.pinot.query.service.dispatch.QueryDispatcher.submitAndReduce(QueryDispatcher.java:117)\norg.apache.pinot.broker.requesthandler.MultiStageBrokerRequestHandler.handleRequest(MultiStageBrokerRequestHandler.java:227)\norg.apache.pinot.broker.requesthandler.BaseBrokerRequestHandler.handleRequest(BaseBrokerRequestHandler.java:168)\n"}],"numGroupsLimitReached":false,"timeUsedMs":0,"requestId":"2087751583000000048","brokerId":"Broker_localhost_21000","numDocsScanned":0,"totalDocs":0,"numEntriesScannedInFilter":0,"numEntriesScannedPostFilter":0,"numServersQueried":0,"numServersResponded":0,"numSegmentsQueried":0,"numSegmentsProcessed":0,"numSegmentsMatched":0,"numConsumingSegmentsQueried":0,"numConsumingSegmentsProcessed":0,"numConsumingSegmentsMatched":0,"minConsumingFreshnessTimeMs":0,"numSegmentsPrunedByBroker":0,"numSegmentsPrunedByServer":0,"numSegmentsPrunedInvalid":0,"numSegmentsPrunedByLimit":0,"numSegmentsPrunedByValue":0,"brokerReduceTimeMs":0,"offlineThreadCpuTimeNs":0,"realtimeThreadCpuTimeNs":0,"offlineSystemActivitiesCpuTimeNs":0,"realtimeSystemActivitiesCpuTimeNs":0,"offlineResponseSerializationCpuTimeNs":0,"realtimeResponseSerializationCpuTimeNs":0,"offlineTotalCpuTimeNs":0,"realtimeTotalCpuTimeNs":0,"explainPlanNumEmptyFilterSegments":0,"explainPlanNumMatchAllFilterSegments":0,"traceInfo":{}}
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQueryInternal(ClusterIntegrationTestUtils.java:733)
	at org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.testQuery(ClusterIntegrationTestUtils.java:699)
	... 34 more
```

**Comment by Jackie-Jiang** [comment]
@gortiz Can you please take a look? Find it in this CI run: https://github.com/apache/pinot/actions/runs/11663662104/job/32472561354?pr=14368

**Comment by gortiz** [comment]
The pretty print stack trace:
```
Caught exception while processing blocks for mailbox: 2087751583000000048|1|0|0|0
java.lang.RuntimeException: Caught exception while processing blocks for mailbox: 2087751583000000048|1|0|0|0
  at org.apache.pinot.query.mailbox.channel.MailboxContentObserver.onNext(MailboxContentObserver.java:95)
  at org.apache.pinot.query.mailbox.channel.MailboxContentObserver.onNext(MailboxContentObserver.java:41)
  at io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onMessage(ServerCalls.java:262)
  at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.messagesAvailableInternal(ServerCallImpl.java:334)
...
Caused by: java.io.IOException: Failed to deserialize data block with serde class org.apache.pinot.common.datablock.ZeroCopyDataBlockSerde
  at org.apache.pinot.common.datablock.DataBlockUtils.deserialize(DataBlockUtils.java:251)
  at org.apache.pinot.common.datablock.DataBlockUtils.readFrom(DataBlockUtils.java:179)
  at org.apache.pinot.query.mailbox.ReceivingMailbox.offerRaw(ReceivingMailbox.java:108)
  at org.apache.pinot.query.mailbox.channel.MailboxContentObserver.onNext(MailboxContentObserver.java:63)
...
Caused by: java.lang.IndexOutOfBoundsException
  at java.base/java.nio.Buffer$1.apply(Buffer.java:757)
  at java.base/java.nio.Buffer$1.apply(Buffer.java:754)
  at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:213)
  at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:210)}
org.apache.pinot.query.service.dispatch.QueryDispatcher.runReducer(QueryDispatcher.java:482)
org.apache.pinot.query.service.dispatch.QueryDispatcher.submitAndReduce(QueryDispatcher.java:117)
org.apache.pinot.broker.requesthandler.MultiStageBrokerRequestHandler.handleRequest(MultiStageBrokerRequestHandler.java:227)
org.apache.pinot.broker.requesthandler.BaseBrokerRequestHandler.handleRequest(BaseBrokerRequestHandler.java:168)
"
```

**Comment by gortiz** [comment]
This seem to be the actual trace:
```
2024-11-04T12:01:23.9656601Z Caused by: java.lang.IndexOutOfBoundsException
2024-11-04T12:01:23.9657421Z 	at java.base/java.nio.Buffer$1.apply(Buffer.java:757) ~[?:?]
2024-11-04T12:01:23.9658309Z 	at java.base/java.nio.Buffer$1.apply(Buffer.java:754) ~[?:?]
2024-11-04T12:01:23.9659341Z 	at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:213) ~[?:?]
2024-11-04T12:01:23.9660535Z 	at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:210) ~[?:?]
2024-11-04T12:01:23.9661755Z 	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:98) ~[?:?]
2024-11-04T12:01:23.9663112Z 	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106) ~[?:?]
2024-11-04T12:01:23.9669347Z 	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302) ~[?:?]
2024-11-04T12:01:23.9670609Z 	at java.base/java.nio.Buffer.checkIndex(Buffer.java:779) ~[?:?]
2024-11-04T12:01:23.9671701Z 	at java.base/java.nio.HeapByteBuffer.getInt(HeapByteBuffer.java:502) ~[?:?]
2024-11-04T12:01:23.9674033Z 	at org.apache.pinot.segment.spi.memory.PinotByteBuffer.getInt(PinotByteBuffer.java:153) ~[pinot-segment-spi-1.3.0-SNAPSHOT.jar:1.3.0-SNAPSHOT-1b7a6ca824c095f8e58eb183118adc647e532676]
2024-11-04T12:01:23.9677625Z 	at org.apache.pinot.common.datablock.ZeroCopyDataBlockSerde.calculateEndOffset(ZeroCopyDataBlockSerde.java:242) ~[pinot-common-1.3.0-SNAPSHOT.jar:1.3.0-SNAPSHOT-1b7a6ca824c095f8e58eb183118adc647e532676]
2024-11-04T12:01:23.9681276Z 	at org.apache.pinot.common.datablock.ZeroCopyDataBlockSerde.deserialize(ZeroCopyDataBlockSerde.java:188) ~[pinot-common-1.3.0-SNAPSHOT.jar:1.3.0-SNAPSHOT-1b7a6ca824c095f8e58eb183118adc647e532676]
2024-11-04T12:01:23.9685304Z 	at org.apache.pinot.common.datablock.DataBlockUtils.deserialize(DataBlockUtils.java:249) ~[pinot-common-1.3.0-SNAPSHOT.jar:1.3.0-SNAPSHOT-1b7a6ca824c095f8e58eb183118adc647e532676]
```

**Comment by gortiz** [comment]
Given the criticality of this part of the code I've been trying to replicate this error for several hours.

They only way I was able to replicate something similar was if the message is so huge that some of the offset overflow.

Specifically I was able to replicate the following trace:
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 4
	at org.apache.pinot.segment.spi.memory.CompoundDataBuffer.getInt(CompoundDataBuffer.java:360)
	at org.apache.pinot.common.datablock.ZeroCopyDataBlockSerde.calculateEndOffset(ZeroCopyDataBlockSerde.java:246)
	at org.apache.pinot.common.datablock.ZeroCopyDataBlockSerde.deserialize(ZeroCopyDataBlockSerde.java:192)
	at org.apache.pinot.common.datablock.ZeroCopyDataBlockSerde.verifyDeserialization(ZeroCopyDataBlockSerde.java:345)
	at org.apache.pinot.common.datablock.ZeroCopyDataBlockSerde.serialize(ZeroCopyDataBlockSerde.java:104)
	at org.apache.pinot.common.datablock.DataBlockUtils.serialize(DataBlockUtils.java:156)
	at org.apache.pinot.common.datablock.DataBlockUtils.serialize(DataBlockUtils.java:141)
	at org.apache.pinot.common.datablock.ZeroCopyDataBlockSerdeTest.testSerde(ZeroCopyDataBlockSerdeTest.java:142)
```

Which fails in the same place but the top of the stack is probably different because the code is not JIT compiled. Also lines may be different because I've added a bunch of new tests while trying to reproduce it.

I'm going to create a PR adding some tests that fail with better messages in this scenario so if this error happens again (and this is the reason) we should be able to detect it sooner.

**Comment by gortiz** [comment]
I'm dumb. Just after writing the previous message I've found that the original error included the query that fails. With that query I was able to reproduce the issue in a very simple way. It looks like that query found a specific combination of dictionary sizes that, by chance ended up finding a bug in PagedPinotOutputStream.

https://github.com/apache/pinot/pull/14389 should fix it. That PR includes two tests: One unit that test this case in PagedPinotOutputStream and an integration that tests the executes the problematic query in the multi stage query engine. The later may be useful to verify this issue is fixed but also as a regression test.
